File: engine/src/com/magma/engine/entities/Entity.java
package com.magma.engine.entities;

import com.badlogic.gdx.Gdx;

public class Entity {
	
	private int health;
	private int energy;
	private int fear;
	
	public final StatusEffect stats;
	
	// makes an immortal entity
	public Entity() {
		this.stats = new StatusEffect();
	}

    public static Entity createImmortal(){
        Entity e = new Entity();
		e.addStatusEffect(StatusEffect.invincible());
		e.replenish();
        return e;
    }

    public static Entity createAdvanced(StatusEffect effect){
        Entity e = new Entity();
		e.addStatusEffect(effect);
		e.replenish();
        return e;
    }

    // spacegame specific stuff, don't worry about this
    public static Entity createSpace(int health, int energy, int defense){
        StatusEffect effect = new StatusEffect();
        effect.defense = defense;
        effect.maxHealth = health;
        effect.maxEnergy= energy;
        effect.maxFear = 100;

        Entity e = new Entity();
		e.addStatusEffect(effect);
		e.replenish();
        return e;
    }
	
	public Entity(StatusEffect stats) {
		this.stats = new StatusEffect();
		
		addStatusEffect(stats);
		replenish();
		
		if (stats.speed == 0) {
			Gdx.app.log("Entity", "NOTE: Entities " + getClass().getSimpleName() + " speed's is zero, so it will not move");
		}
	}
	
	public void addStatusEffect(StatusEffect stat) {
		stats.add(stat);
	}
	
	public void removeStatusEffect(StatusEffect stat) {
		stats.subtract(stat);
        //TODO: add death
	}
	
    public void damage(int amount){
        health -= amount;
    }

	public void replenish() {
		this.health = stats.maxHealth;
		this.energy = stats.maxEnergy;
		this.fear = 0;
	}
	
	public StatusEffect getStats() {
		return stats;
	}
	
	public String toString() {
		return "HP: " + health + "\nEN: " + energy + "\nFE: " + fear + "\n" + getStats().toString();
	}

	public int getHealth() {
		return health;
	}

	public void setHealth(int health) {
		this.health = health;
	}

	public int getEnergy() {
		return energy;
	}

	public void setEnergy(int energy) {
		this.energy = energy;
	}

	public int getFear() {
		return fear;
	}

	public void setFear(int fear) {
		this.fear = fear;
	}
}

File: engine/src/com/magma/engine/entities/Entity.kt
package com.magma.engine.entities

import com.badlogic.gdx.Gdx

open class Entity {
    var health = 0
    var energy = 0
    var fear = 0
    val stats: StatusEffect

    // makes an immortal entity
    constructor() {
        stats = StatusEffect()
    }

    constructor(stats: StatusEffect) {
        this.stats = StatusEffect()
        addStatusEffect(stats)
        replenish()
        if (stats.speed == 0) {
            Gdx.app.log("Entity", "NOTE: Entities " + javaClass.simpleName + " speed's is zero, so it will not move")
        }
    }

    fun addStatusEffect(stat: StatusEffect) {
        stats.add(stat)
    }

    fun removeStatusEffect(stat: StatusEffect) {
        stats.subtract(stat)
        //TODO: add death
    }

    fun damage(amount: Int) {
        health -= amount
    }

    fun replenish() {
        health = stats.maxHealth
        energy = stats.maxEnergy
        fear = 0
    }

    override fun toString(): String {
        return "${this.javaClass.simpleName} | HP: $health | EN: $energy | FE: $fear | $stats"
    }

    companion object {
        val immortal : Entity
            get() {
                val e = Entity()
                e.addStatusEffect(StatusEffect.invincible)
                e.replenish()
                return e
            }

        fun createAdvanced(effect: StatusEffect): Entity {
            val e = Entity()
            e.addStatusEffect(effect)
            e.replenish()
            return e
        }

        // spacegame specific stuff, don't worry about this
        fun createSpace(health: Int, energy: Int, defense: Int): Entity {
            val effect = StatusEffect()
            effect.defense = defense
            effect.maxHealth = health
            effect.maxEnergy = energy
            effect.maxFear = 100
            val e = Entity()
            e.addStatusEffect(effect)
            e.replenish()
            return e
        }
    }
}